x-pack/filebeat/input/websocket/input.go

```diff
@@ -0,0 +1,381 @@
+
+				return
+			}
+			metrics.receivedBytesTotal.Add(uint64(len(message)))
+			state["response"] = message
+			log.Debugw("received websocket message", logp.Namespace("websocket"), string(message))
+			err = i.processAndPublishData(ctx, metrics, prg, ast, state, cursor, pub, log)
+			if err != nil {
+				metrics.errorsTotal.Inc()
+				log.Errorw("failed to process and publish data", "error", err)
+				return
+			}
+		}
+	}()
+
+	// blocks until done is closed or context is cancelled
+	for {
+		select {
+		case <-done:
+			return nil
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// processAndPublishData processes the data in state, updates the cursor and publishes it to the publisher.
+// the CEL program here only executes a single time, since the websocket connection is persistent and events are received and processed in real time.
+func (i *input) processAndPublishData(ctx context.Context, metrics *inputMetrics, prg cel.Program, ast *cel.Ast,
+	state map[string]interface{}, cursor map[string]interface{}, pub inputcursor.Publisher, log *logp.Logger) error {
+	goodCursor := cursor
+	start := i.now().In(time.UTC)
+	state, err := evalWith(ctx, prg, ast, state, start)
+	log.Debugw("cel engine eval state", logp.Namespace("websocket"), "state", redactor{state: state, cfg: i.cfg.Redact})
+	if err != nil {
+		metrics.celEvalErrors.Add(1)
+		switch {
+		case errors.Is(err, context.Canceled), errors.Is(err, context.DeadlineExceeded):
+			return err
+		default:
+			metrics.errorsTotal.Inc()
+		}
+		log.Errorw("failed evaluation", "error", err)
+	}
+	metrics.celProcessingTime.Update(time.Since(start).Nanoseconds())
+
+	e, ok := state["events"]
+	if !ok {
+		log.Errorw("unexpected missing events from evaluation")
+	}
+	var events []interface{}
+	switch e := e.(type) {
+	case []interface{}:
+		if len(e) == 0 {
+			return nil
+		}
+		events = e
+	case map[string]interface{}:
+		if e == nil {
+			return nil
+		}
+		log.Debugw("single event object returned by evaluation", "event", e)
+		events = []interface{}{e}
+	default:
+		return fmt.Errorf("unexpected type returned for evaluation events: %T", e)
+	}
+
+	metrics.eventsReceived.Add(uint64(len(events)))
+
+	// Drop events from state. If we fail during the publication,
+	// we will reprocess these events.
+	delete(state, "events")
+
+	// Get cursors if they exist.
+	var (
+		cursors      []interface{}
+		singleCursor bool
+	)
+	if c, ok := state["cursor"]; ok {
+		cursors, ok = c.([]interface{})
+		if ok {
+			if len(cursors) != len(events) {
+				log.Errorw("unexpected cursor list length", "cursors", len(cursors), "events", len(events))
+				// But try to continue.
+				if len(cursors) < len(events) {
+					cursors = nil
+				}
+			}
+		} else {
+			cursors = []interface{}{c}
+			singleCursor = true
+		}
+	}
+	// Drop old cursor from state. This will be replaced with
+	// the current cursor object below; it is an array now.
+	delete(state, "cursor")
+
+	//start = time.Now()
```

Comment:
?

Reply:
Was thinking about keeping batch processing time, but here usually a batch will be a single event in most scenarios.

Reply:
Yes, I think we need this for metrics.